### ISSUE TITLE
Migrate to/from file

### ISSUE THREAD

**Issue by abernh** [CodingButter-LLC/directus-migrator#7]
Via the discussions here https://github.com/directus/directus/discussions/17495 and there https://github.com/directus/directus/discussions/13041 I came by your migrator script.

Very clean setup, nice code. Impressive work. (just missing tests ;) )

My current focus is on solving the issue of deploying Directus instances including all the migration-information. I want to "save" the current state (schema, permissions, roles, etc) into the repository and on `postinstall` a "migrate-like" script is run to import all that.

Your script does already all the heavy lifting by retrieving the contents from a source instance, compare it to the target and imports it. It is also nicely split into specific migrator controllers (and therefore well extendable) and I'd like to suggest to extend it to allow e.g. `environments` not only to point to a server via `{name, endpoint, accessToken}` but to local files `{name, filename}`, e.g.

```js
// directus-migrator.config.js
...
    {
      "name": "repo",
      "filename": "./current.directus-migrator.json"
    },
```

This would make it possible to 
 - export the current environment state into a single "migration file", commit to the repo and
 - import the current environment state on "postinstall" from the repo state

Let me know what you think about it.
And yes, I'd like to collaborate on that - I'm already working on a first version as I really need that functionality :)

**Comment by CodingButter** [comment]
Thanks for your interest in the project! This is my first project that has gained any real traction. If you'd be willing to help that would be appreciated.

**Comment by CodingButter** [comment]
> Via the discussions here [directus/directus#17495](https://github.com/directus/directus/discussions/17495) and there [directus/directus#13041](https://github.com/directus/directus/discussions/13041) I came by your migrator script.
> 
> Very clean setup, nice code. Impressive work. (just missing tests ;) )
> 
> My current focus is on solving the issue of deploying Directus instances including all the migration-information. I want to "save" the current state (schema, permissions, roles, etc) into the repository and on `postinstall` a "migrate-like" script is run to import all that.
> 
> Your script does already all the heavy lifting by retrieving the contents from a source instance, compare it to the target and import it. It is also nicely split into specific migrator controllers (and therefore well extendable) and I'd like to suggest to extend it to allow `environments` not only to point to a server via `{name, endpoint, accessToken}` but also to local files `{name, filename}`, e.g.
> 
> ```js
> // directus-migrator.config.js
> ...
>     {
>       "name": "repo",
>       "filename": "./current.directus-migrator.json"
>     },
> ```
> 
> This would make it possible to
> 
> * export the current environment state into a single "migration file", commit to the repo and
> * import the current environment state on "postinstall" from the repo state
> 
> Let me know what you think about it. And yes, I'd like to collaborate on that - I'm already working on a first version as I really need that functionality :)

are you suggesting a json file with the schema, permissions, roles ... and use that to migrate instead of directus environment?

**Comment by xu4wang** [comment]
If possible, I would suggest a JSON file for each role, and a JSON file for each flow.

The permissions for the corresponding role should be inside the role definition JSON.
The operations for the corresponding flow should be inside the flow definition JSON.

In this way, the people working on different flows can work separately, and the flow configuration along with the operations can be exported and version controlled.

**Comment by CodingButter** [comment]
so you could essentially migrate to and from the json files

**Comment by xu4wang** [comment]
> so you could essentially migrate to and from the json files

Yes..  I was thinking to build it myself and created a repo https://github.com/xu4wang/directools a few days ago.
It will be great if directus-migrator can support this..

**Comment by xu4wang** [comment]
Or if we can put a path in the environment to migrate to/from files.  for example
target = { name: "file-system", endpoint: "file://./", accessToken: '' }
